pkg/util/pki/parse.go

```diff
@@ -140,3 +140,160 @@ func DecodeX509CertificateRequestBytes(csrBytes []byte) (*x509.CertificateReques
+	// Build the nodes to the linked list
+	for i := range certs {
+		chains = append(chains, &chainNode{cert: certs[i]})
+	}
+
+	for {
+		// We have built the full chain
+		if len(chains) == 1 {
+			break
+		}
+
+		// lastChainsLength is used to ensure that at every pass, the number of
+		// tested chains gets smaller.
+		lastChainsLength := len(chains)
+		for i := 0; i < len(chains)-1; i++ {
+			for j := 1; j < len(chains); j++ {
+				if i == j {
+					continue
+				}
+
+				// attempt to add both chain together
+				chain, ok := chains[i].tryMergeChain(chains[j])
+				if ok {
+					// If adding the chains together was successful, remove inner chain from
+					// list
+					chains = append(chains[:j], chains[j+1:]...)
+				}
+
+				chains[i] = chain
+			}
+		}
+
+		// If no chains were merged in this pass, the chain can never be built
+		// flat. Error.
+		if lastChainsLength == len(chains) {
+			return nil, errors.NewInvalidData("certificate chain is malformed or broken")
+		}
+	}
+
+	// There is only a single chain left at index 0. Return chain as PEM.
+	return chains[0].toBundleAndCA()
+}
+
+// toBundleAndCA will return the PEM bundle of this chain.
+func (c *chainNode) toBundleAndCA() (*PEMBundle, error) {
+	var (
+		certs []*x509.Certificate
+		ca    *x509.Certificate
+	)
+
+	for {
+		// If the issuer is nil, we have hit the root of the chain. Assign the CA
+		// to this certificate and stop traversing.
+		if c.issuer == nil {
+			ca = c.cert
+			break
+		}
+
+		// Add this nodes certificate to the list at the end. Ready to check next
+		// node up.
+		certs = append(certs, c.cert)
+		c = c.issuer
+	}
+
+	capem, err := EncodeX509(ca)
+	if err != nil {
+		return nil, err
+	}
+
+	// If no certificates parsed, then CA is the only certificate and should be
+	// the chain
+	if len(certs) == 0 {
+		return &PEMBundle{ChainPEM: capem}, nil
+	}
+
+	// Encode cert chain
+	chainpem, err := EncodeX509Chain(certs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return chain and ca
+	return &PEMBundle{CAPEM: capem, ChainPEM: chainpem}, nil
+}
+
+// tryMergeChain will attempt to add the chain "c" and "chain" together.
+// Successful if the either chain is added to the top or bottom of the other
+// chain.
+func (c *chainNode) tryMergeChain(chain *chainNode) (*chainNode, bool) {
+	// The given chain's root has been signed by this node. Add this node on top
+	// of the given chain.
+	if chain.root().cert.CheckSignatureFrom(c.cert) == nil {
+		chain.root().issuer = c
+		return chain, true
+	}
+
+	// The given chain is the issuer of the root of this node. Add the given
+	// chain on top of the root of this node.
+	if c.root().cert.CheckSignatureFrom(chain.cert) == nil {
+		root := c.root()
+		root.issuer = chain
+		return c, true
+	}
+
+	// Chains cannot be added together.
+	return c, false
+}
+
+// Return the root most node of this chain.
+func (c *chainNode) root() *chainNode {
+	if c.issuer == nil {
+		return c
+	}
+	return c.issuer.root()
```

Comment:
extreme nit: I assume this will never happen in practice but a really long chain would cause a DDoS here as you will smash the stack: https://godbolt.org/z/r4b9Y1xMs

```suggestion
	for c.issuer != nil {
		c = c.issuer
	}
	return c
```

Reply:
Ah sorry, I'm a dumb dumb. Didn't see the for loop :facepalm: 

Yeah, that's probably better tbh

Reply:
@jakexks I just looked at the snippet you wrote (godbolt.org/z/r4b9Y1xMs) and WOAHH I didn't know there existed a "go playground" where we can immediately see the disassembly!!! Thanks!